lib/kubernetes-deploy/deploy_task.rb

```diff
@@ -405,6 +379,41 @@ def log_pruning(kubectl_output)
       @logger.summary.add_action("pruned #{pruned.length} #{'resource'.pluralize(pruned.length)}")
     end
 
+    def record_apply_failure(err)
+      warn_msg = "WARNING: Any resources not mentioned in the error(s) below were likely created/updated. " \
+        "You may wish to roll back this deploy."
+      @logger.summary.add_paragraph(ColorizedString.new(warn_msg).yellow)
+
+      unidentified_errors = []
+      err.each_line do |line|
+        bad_files = find_bad_files_from_kubectl_output(line)
+        if bad_files.present?
+          bad_files.each { |f| record_invalid_template(err: f[:err], filename: f[:filename], content: f[:content]) }
+        else
+          unidentified_errors << line
+        end
+      end
+
+      if unidentified_errors.present?
+        msg = "#{ColorizedString.new('Unidentified error(s):').red}\n#{indent_four(unidentified_errors.join)}"
+        @logger.summary.add_paragraph(msg)
+      end
+    end
+
+    # Inspect the file referenced in the kubectl stderr
+    # to make it easier for developer to understand what's going on
+    def find_bad_files_from_kubectl_output(line)
+      # stderr often contains one or more lines like the following, from which we can extract the file path(s):
+      # Error from server (TypeOfError): error when creating "/path/to/service-gqq5oh.yml": Service "web" is invalid:
+
+      line.scan(%r{"(/\S+\.ya?ml\S*)"}).each_with_object([]) do |matches, bad_files|
+        matches.each do |path|
+          content = File.read(path) if File.file?(path)
+          bad_files << { filename: File.basename(path), err: line, content: content }
```

Comment:
This lets us split up the various errors we've identified appropriately.

Example before this commit:

![image](https://user-images.githubusercontent.com/4789493/38000976-51a6e6a8-31f8-11e8-89e1-37dbb16c63a4.png)


Example after:

![image](https://user-images.githubusercontent.com/4789493/38000937-2a3ce996-31f8-11e8-8304-46acbfd21397.png)